pkg/node/discovery.go

```diff
@@ -0,0 +1,175 @@
+// This file is part of MinIO Direct CSI
+// Copyright (c) 2020 MinIO, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package node
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	directv1alpha1 "github.com/minio/direct-csi/pkg/apis/direct.csi.min.io/v1alpha1"
+	"github.com/minio/direct-csi/pkg/dev"
+
+	"github.com/golang/glog"
+	simd "github.com/minio/sha256-simd"
+)
+
+func findDrives(ctx context.Context, nodeID string, procfs string) ([]directv1alpha1.DirectCSIDrive, error) {
+	drives := []directv1alpha1.DirectCSIDrive{}
+	devs, err := dev.FindDevices(ctx)
+	if err != nil {
+		return drives, err
+	}
+
+	for _, d := range devs {
+		partitions := d.GetPartitions()
+		if len(partitions) > 0 {
+			for _, partition := range partitions {
+				drive, err := makePartitionDrive(nodeID, partition, d.Devname)
+				if err != nil {
+					glog.Errorf("Error discovering parition %s: %v", d.Devname, err)
+					continue
+				}
+				drives = append(drives, *drive)
+			}
+			continue
+		}
+
+		drive, err := makeRootDrive(nodeID, d)
+		if err != nil {
+			return nil, err
+		}
+		drives = append(drives, *drive)
+	}
+
+	return drives, nil
+}
+
+func makePartitionDrive(nodeID string, partition dev.Partition, rootPartition string) (*directv1alpha1.DirectCSIDrive, error) {
```

Comment:
The `makePartitionDrive` and `makeRootDrive` are pretty much using the same code. I can send a PR after this to unify it.